data_warehouse/services/coverage_statistics_service.py

```diff
@@ -0,0 +1,202 @@
+            coverage_count=Count('technical_title'))
+        for item in coverage_titles:
+            data.append(
+                {
+                    'title': item['technical_title'],
+                    'coverage_count': item['coverage_count'],
+                    'total_count': User.objects.filter(
+                        technical_title=item['technical_title']).count()
+                }
+            )
+        return data
+
+    @staticmethod
+    def groupby_departments(users_qs):
+        '''按照部门分组
+
+        Parameters
+        -------
+        users_qs: QuerySet
+            待分组的User查询集
+
+        Returns
+        ------
+        list of dict
+            按部门后分组后的结果，list包含的元素dict定义为： {
+                department: string,
+                coverage_count: int,
+                    参数users_qs包含的用户中，隶属该部门的用户人数。
+                total_count: int
+                    系统里隶属该部门的用户总人数。
+            }
+        '''
+        data = []
+        coverage_departments = users_qs.values(
+            'administrative_department__name').annotate(coverage_count=Count(
+                'administrative_department__name'))
+        for item in coverage_departments:
+            data.append(
+                {
+                    'department': item['administrative_department__name'],
+                    'coverage_count': item['coverage_count'],
+                    'total_count': User.objects.filter(
+                        department__name=item[
+                            'administrative_department__name']).count()
+                }
+            )
+        return data
+
+    @staticmethod
+    def get_traning_records(user, program_id=None, department_id=None,
+                            start_time=None, end_time=None):
+        '''根据项目ID和起始时间查询所有的培训活动的所有的培训记录（包含校内和校外活动)
+
+        Parameters
+        ----
+        user: User
+            执行此操作的用户
+
+        program_id: int
+            培训项目ID，为None表示查询所有的培训项目。
+
+        department_id: int
+            需要导出培训项目所属的部门，非校级管理员必须给此参数，校级管理员为此参数None则导出所有院系的所有培训项目。
+
+        start_time: datetime
+            培训活动开始时间
+
+        end_time: datetime
+            培训活动结束时间
+
+        Return
+        ------
+
+        QuerySet
+            包含Record的查询集
+        '''
+        if user is None or not (user.is_department_admin or user
+                                .is_school_admin):
+            raise BadRequest('你不是管理员，无权操作。')
+
+        if department_id is not None:
+            department = Department.objects.filter(id=department_id)
+            if not department:
+                raise BadRequest('部门不存在。')
+            if not (user.is_school_admin or user.check_department_admin(
+                    department[0])):
+                raise BadRequest('你不是该院系的管理员，无权操作。')
+            department_ids = [department_id]
+        else:
+            if user.is_school_admin:
+                department_ids = DepartmentService.get_top_level_departments(
```

Comment:
```python
department_ids = (
    DepartmentService.get_top_level_departments()
    .values_list('id', flat=True)
)
```

Reply:
> ```python
> department_ids = (
>     DepartmentService.get_top_level_departments()
>     .values_list('id', flat=True)
> )
> ```

Why we need to add `()`?

Reply:
To satisfy the PEP-8 style while formatting the code in a clean way.

Reply:
🤔️ the code snippet has passed the check of pep-8.

Reply:
```python
# 1
department_ids = (
    DepartmentService.get_top_level_departments()
    .values_list('id', flat=True)
)
# 2
department_ids = DepartmentService.get_top_level_departments(
                ).values_list('id', flat=True)
```
It it, but don't you think the former looks better?

Reply:
it' s up to you. 😄